zebrad/src/components/sync.rs

```diff
@@ -584,65 +518,16 @@ where
     async fn request_blocks(&mut self, hashes: Vec<block::Hash>) -> Result<(), Report> {
         tracing::debug!(hashes.len = hashes.len(), "requesting blocks");
         for hash in hashes.into_iter() {
-            // Avoid re-downloading blocks that have already been verified.
-            // This is particularly important for nodes on slow or unreliable
-            // networks.
+            // If we've queued the download of a hash behind our current chain tip,
+            // we've been given bad responses by our peers.  Abort the sync and restart.
             if self.state_contains(hash).await? {
-                tracing::debug!(
-                    ?hash,
-                    "request_blocks: Unexpected duplicate hash: already in state"
-                );
-                continue;
+                return Err(eyre!("queued download of hash behind our chain tip"));
```

Comment:
Consider using ? rather than return - it provides better error context.
```suggestion
                Err(eyre!("queued download of hash behind our chain tip"))?;
```

Reply:
Hmm, adding this change made clippy unhappy.

Reply:
yes, it do that, ignore clippy on this one it's wrong. I have been meaning to disable that lint for ages with the intent of eventually introducing the opposite lint.

Reply:
Weird, I thought we'd added allow(try_err) a few weeks ago

Reply:
Hmm, apparently not, it's not in lib.rs.

Reply:
Let's wait until there's a lint one way or the other, then use it to ensure consistency.

Reply:
I'm confused. Do you think the existing clippy lint is good advice?
Or do you think that consistency with all the default linter rules is really important?

I suggested that we use the ? operator, because it provides better error context.

I believe that benefit outweighs the cost of any minor inconsistencies that might happen from disabling a single lint.

(If we wait for a changed lint to hit clippy stable, we might be waiting months. In the meantime, our error reporting will be worse.)

Reply:
I agree that it's a bad idea to have the lint on for some crates, and off for others.

I can submit a PR that turns it off everywhere - it's already disabled in most of the large zebra crates.

Reply:
I don't think that it's particularly important either way, and in this case I think that it makes no meaningful difference at all. I'd be happy using one, or the other. But I would prefer that if we want to make a policy choice, we do that in a consistent way, independent of this PR.

Reply:
Ok, this is now ticket #1199.